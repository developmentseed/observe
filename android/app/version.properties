VERSION_CODE=14
