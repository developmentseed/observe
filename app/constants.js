export const EDIT_PENDING_STATUS = 'EDIT_PENDING_STATUS'
export const EDIT_UPLOADING_STATUS = 'EDIT_UPLOADING_STATUS'
export const EDIT_UPLOADED_STATUS = 'EDIT_UPLOADED_STATUS'
export const EDIT_SUCCEEDED_STATUS = 'EDIT_SUCCEEDED_STATUS'

export const PHOTO_PENDING_STATUS = 'PHOTO_PENDING_STATUS'
export const PHOTO_PENDING_EDIT_STATUS = 'PHOTO_PENDING_EDIT_STATUS'
export const PHOTO_UPLOADING_STATUS = 'PHOTO_UPLOADING_STATUS'
export const PHOTO_UPLOADED_STATUS = 'PHOTO_UPLOADED_STATUS'
export const PHOTO_SUCCEEDED_STATUS = 'PHOTO_SUCCEEDED_STATUS'

export const TRACE_PENDING_STATUS = 'TRACE_PENDING_STATUS'
export const TRACE_PENDING_EDIT_STATUS = 'TRACE_PENDING_EDIT_STATUS'
export const TRACE_UPLOADING_STATUS = 'TRACE_UPLOADING_STATUS'
export const TRACE_UPLOADED_STATUS = 'TRACE_UPLOADED_STATUS'
export const TRACE_SUCCEEDED_STATUS = 'TRACE_SUCCEEDED_STATUS'
