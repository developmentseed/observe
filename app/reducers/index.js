import { reducer as network } from 'react-native-offline'
import { combineReducers } from 'redux'
import { persistReducer } from 'redux-persist'
import storage from 'redux-persist/lib/storage'
import FilesystemStorage from 'redux-persist-filesystem-storage'

import AboutReducer from './about'
import AuthorizationReducer from './authorization'
import MapReducer from './map'
import PresetReducer from './presets'
import AccountReducer from './account'
import EditReducer from './edit'
import NotificationReducer from './notification'
import CameraReducer from './camera'
import TracesReducer from './traces'
import ObserveAPIReducer from './observeApi'
import CurrentWayEditReducer from './currentWayEdit'

const authorizationPersistConfig = {
  key: 'authorization',
  storage,
  whitelist: ['isAuthorized']
}

const mapPersistConfig = {
  key: 'map',
  storage: FilesystemStorage,
  whitelist: ['fetchedTiles', 'lru', 'offlineResources', 'pendingEviction']
}

const editPersistConfig = {
  key: 'edit',
  storage,
  whitelist: [
    'edits',
    'uploadedEdits',
    'editsGeojson'
  ]
}

const tracesPersistConfig = {
  key: 'traces',
  storage,
  whitelist: [
    'currentTrace',
    'traces'
  ]
}

const observeApiPersistConfig = {
  key: 'observeApi',
  storage,
  whitelist: [
    'token'
  ]
}

const photosPersistConfig = {
  key: 'photos',
  storage,
  whitelist: [
    'photos'
  ]
}

const rootReducer = combineReducers({
  about: AboutReducer,
  authorization: persistReducer(authorizationPersistConfig, AuthorizationReducer),
  map: persistReducer(mapPersistConfig, MapReducer),
  presets: PresetReducer,
  account: AccountReducer,
  edit: persistReducer(editPersistConfig, EditReducer),
  currentWayEdit: CurrentWayEditReducer,
  network,
  notification: NotificationReducer,
  traces: persistReducer(tracesPersistConfig, TracesReducer),
  observeApi: persistReducer(observeApiPersistConfig, ObserveAPIReducer),
  photos: persistReducer(photosPersistConfig, CameraReducer)
})

export default rootReducer
