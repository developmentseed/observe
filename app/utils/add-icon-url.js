import * as taginfo from '../presets/taginfo.json'
import { uninterestingKeys } from './uninterestingKeys'
import _omit from 'lodash.omit'
import _find from 'lodash.find'

/**
 * Parses through each feature, and adds icon url to properties
 * The icon property is used for styling later
 * @param {Object} geojson
 * @return {Object} geojson
 */
export function addIconUrl (geojson) {
  if (geojson.type === 'FeatureCollection') {
    const points = geojson.features.filter(feature => {
      return feature.geometry.type === 'Point'
    })
    points.forEach(feature => {
      setIcon(feature)
    })
  } else {
    setIcon(geojson)
  }
  return geojson
}

function setIcon (feature) {
  feature.properties.icon = undefined
  const featureTags = _omit(feature.properties, uninterestingKeys)
  const keys = Object.keys(featureTags)
  for (const index in keys) {
    const key = keys[index]
    const match = _find(taginfo.tags, { 'key': key, 'value': featureTags[key] })
    if (match) {
      if (match.icon_url != null) {
        feature.properties.icon = match.icon_url.replace(/-/g, '_')
      }
      break
    }
  }
  if (!feature.properties.icon || feature.properties.icon.startsWith('iD-')) {
    feature.properties.icon = 'maki_marker'
  }
  return feature
}
