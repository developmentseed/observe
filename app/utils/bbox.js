import tileCover from '@mapbox/tile-cover'
import bboxPolygon from '@turf/bbox-polygon'
import turfBbox from '@turf/bbox'
import turfBuffer from '@turf/buffer'
import _flatten from 'lodash.flatten'

// TODO - move this to an appropriate config file
const tileLimits = {
  min_zoom: 16,
  max_zoom: 16
}

/**
 * Convert map bbox object to a string
 * @param {*} bbox
 * @return string
 */
export function bboxToString (bbox) {
  const ne = bbox[0]
  const sw = bbox[1]
  const left = sw[0]
  const bottom = sw[1]
  const right = ne[0]
  const top = ne[1]
  return [left, bottom, right, top].join(',')
}

export function bboxToTiles (bbox) {
  const polygon = bboxPolygon(_flatten(bbox))
  const tiles = tileCover.indexes(polygon.geometry, tileLimits)
  return tiles
}

export function featureToTiles (feature) {
  let bbox = []
  if (feature.geometry && feature.geometry.type === 'Point') {
    const buffer = turfBuffer(feature, '0.0005', { 'units': 'kilometers' })
    bbox = turfBbox(buffer)
  } else {
    bbox = turfBbox(feature)
  }
  const polygon = bboxPolygon(bbox)
  const tiles = tileCover.indexes(polygon.geometry, tileLimits)
  return tiles
}
