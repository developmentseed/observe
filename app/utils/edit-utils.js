import { UploadNetworkError } from './errors'

/**
 * Given an edit, says if the edit can be retried based on the error type
 * @param {Object} edit
 * @returns {Boolean} - true if retryable, false if not
 */
export function isRetryable (edit) {
  return getError(edit).retryable
}

/**
 * For an array of edit objects, returns a filtered list of retriable edits
 * @param {Array<Object>} edits - array of edit objects
 * @returns {Array<Object>}
 */
export function getAllRetriable (edits) {
  return edits.filter(e => isRetryable(e))
}

/**
 * Given an edit, returns whether it has a Version Conflict error
 * @param {Object} edit
 * @returns {Boolean}
 */
export function isConflict (edit) {
  return getError(edit).code === 'VersionMismatchError'
}

/**
 * Return the applicable / current error for an edit.
 * If it has errors, return the last error in the error.
 * If it does not have errors, we assume that it is waiting for network, and return a NetworkError
 * @param {Object} edit - edit object
 * @return {ObserveError} - an instance of ObserveError, with the error for the edit
 */
export function getError (edit) {
  return edit.errors.length > 0 ? edit.errors.slice(-1)[0] : new UploadNetworkError()
}

export function getErrorExtra (edit) {
  return getError(edit).extra
}
