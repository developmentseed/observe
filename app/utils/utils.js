export function isNewId (id) {
  id = String(id)
  return id.startsWith('observe-')
}

export function isWayPendingUpload (feature) {
  const id = feature.properties.id.split('/')[1]
  const ndrefs = feature.properties.ndrefs ? feature.properties.ndrefs.length : undefined

  if (isNewId(id) && ndrefs) {
    return true
  } else {
    return false
  }
}

export function isInvalidFeature (feature) {
  return (
    (feature.geometry.type === 'LineString' && feature.properties.ndrefs.length < 2) ||
    (feature.geometry.type === 'Polygon' && feature.properties.ndrefs.length < 3)
  )
}
