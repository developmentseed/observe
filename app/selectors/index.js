import { createSelector } from 'reselect'
import _flatten from 'lodash.flatten'
import fromEntries from 'fromentries'
import bboxPolygon from '@turf/bbox-polygon'
import booleanIntersects from '@turf/boolean-intersects'
import { featureCollection } from '@turf/helpers'

import { addIconUrl } from '../utils/add-icon-url'
import { bboxToTiles } from '../utils/bbox'
import {
  getTraceLength
} from '../utils/traces'
import cache from '../utils/data-cache'
import { filterTags } from '../utils/filter-tags'

export const getVisibleBounds = state => state.map.visibleBounds
export const getZoom = state => state.map.zoom
const getIntegerZoom = state =>
  state.map.zoom != null ? Math.floor(state.map.zoom) : state.map.zoom

// quantize bounds to prevent excessive filtering when panning around limited
// areas with large numbers of features
const getQuantizedBounds = createSelector(
  getVisibleBounds,
  visibleBounds => {
    if (visibleBounds != null) {
      // visible bounds is [[maxX, maxY], [minX, minY]]
      const width = visibleBounds[0][0] - visibleBounds[1][0]
      const height = visibleBounds[0][1] - visibleBounds[1][1]

      // expand by 50% in each direction
      let minX = visibleBounds[1][0] - width / 2
      let maxX = visibleBounds[0][0] + width / 2
      let minY = visibleBounds[1][1] - height / 2
      let maxY = visibleBounds[0][1] + height / 2

      // truncate to 3 decimal places and round
      minX = Math.floor(minX * 200) / 200
      maxX = Math.ceil(maxX * 200) / 200
      minY = Math.floor(minY * 200) / 200
      maxY = Math.ceil(maxY * 200) / 200

      // serialize as a string to take advantage of reselect equality checks (===)
      return JSON.stringify([[maxX, maxY], [minX, minY]])
    }
  }
)

const getFetchedTiles = state => state.map.fetchedTiles

const getSerialNumber = state => state.map.serialNumber

export const getVisibleFeatures = createSelector(
  [getQuantizedBounds, getFetchedTiles, getIntegerZoom, getSerialNumber],
  (_visibleBounds, fetchedTiles, zoom, serialNumber) => {
    if (_visibleBounds != null && zoom >= 16) {
      const visibleBounds = JSON.parse(_visibleBounds)
      // determine tile coverage
      const tileKeys = bboxToTiles(visibleBounds)

      // load tiled data (where available)
      const tiles = tileKeys.map(k => fetchedTiles[k]).filter(t => t != null)

      // convert bounds to a polygon for intersection purposes
      const bbox = bboxPolygon(_flatten(visibleBounds))

      // track ids of potentially duplicated features (across tile boundaries)
      const ids = new Set()

      // assemble a list of features intersecting the bounding box
      const visibleFeatures = tiles.reduce((features, { key }) => {
        // load data from an LRU cache to prevent needing to store *all* cached
        // data in memory (in addition to persisting it to disk)
        const data = cache.get(key) || {
          features: []
        }

        const newFeatures = data.features
          // ignore duplicates
          .filter(f => !ids.has(f.id))
          // intersect with the bbox
          .filter(f => booleanIntersects(bbox, f))

        // update the list of feature ids
        newFeatures.forEach(f => ids.add(f.id))

        // in-place append
        features.push(...newFeatures)

        return features
      }, [])

      const filtered = filterTags(featureCollection(visibleFeatures))

      return addIconUrl(filtered)
    }

    // no features; empty FeatureCollection
    return featureCollection([])
  }
)

export const isLoadingData = state => state.map.activeTileRequests.length > 0

export const getOfflineResources = ({ map: { offlineResources } }) =>
  offlineResources

export const getOfflineResourceStatus = createSelector(
  getOfflineResources,
  offlineResources =>
    fromEntries(
      Object.entries(offlineResources).map(([name, resource]) => {
        const { data, mapTiles, satelliteTiles, satellite } = resource

        let percentage = (data.percentage + mapTiles.percentage) / 2
        let tilePercentage = mapTiles.percentage
        let size = data.size + mapTiles.size
        let total = mapTiles.total
        let completed = mapTiles.completed

        if (satellite) {
          percentage =
            (data.percentage +
              mapTiles.percentage +
              satelliteTiles.percentage) /
            3
          tilePercentage = (mapTiles.percentage + satelliteTiles.percentage) / 2
          size += satelliteTiles.size

          if (satelliteTiles.total != null) {
            total += satelliteTiles.total
          }

          completed += satelliteTiles.completed
        }

        return [
          name,
          {
            ...resource,
            percentage,
            dataPercentage: data.percentage,
            tilePercentage,
            size,
            total,
            completed
          }
        ]
      })
    )
)

export const getPendingEviction = state => state.map.pendingEviction

export const getIsTracing = state => !!state.traces.currentTrace

export const getCurrentTraceGeoJSON = state => {
  if (state.traces.currentTrace && state.traces.currentTrace.geometry.coordinates.length > 2) {
    return {
      'type': 'FeatureCollection',
      'features': [ state.traces.currentTrace ]
    }
  } else {
    return {
      'type': 'FeatureCollection',
      'features': []
    }
  }
}

export const getCurrentTraceLength = state => {
  const currentTrace = state.traces.currentTrace
  if (!currentTrace || currentTrace.geometry.coordinates.length < 2) return 0
  return getTraceLength(currentTrace)
}

/**
 *
 * @param {Object} state
 * @returns {String} - one of 'recording', 'paused', 'none'
 */
export const getCurrentTraceStatus = state => {
  const { currentTrace, paused } = state.traces
  if (!currentTrace) return 'none'
  if (paused) return 'paused'
  return 'recording'
}

export const showRecordingHeader = state => {
  const { currentTrace, saving } = state.traces
  if (saving) return false
  if (currentTrace) return true
}

export const getTracesGeojson = state => {
  const traces = state.traces.traces
  const tracesGeojson = {
    'type': 'FeatureCollection',
    'features': []
  }
  traces.map(trace => {
    return tracesGeojson.features.push(trace.geojson)
  })

  return tracesGeojson
}

export const getPhotosGeojson = state => {
  const photos = state.photos.photos
  const photosGeojson = {
    'type': 'FeatureCollection',
    'features': []
  }

  if (photos.length > 0) {
    photos.map(photo => {
      const photoGeojson = {
        'type': 'Feature',
        'geometry': {
          'type': 'Point',
          'coordinates': [
            photo.location.coords.longitude,
            photo.location.coords.latitude
          ]
        },
        'properties': {
          'id': photo.id,
          'type': 'photo',
          'description': photo.description,
          'path': photo.path,
          'timestamp': photo.location.timestamp
        }
      }
      return photosGeojson.features.push(photoGeojson)
    })
  }

  return photosGeojson
}

export const getVisibleTiles = state => {
  return bboxToTiles(state.map.visibleBounds)
}
