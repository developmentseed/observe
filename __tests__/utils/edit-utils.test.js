/* global fetch */
/* global test */
/* global __dirname */
/* global expect */

import { getAllRetriable } from '../../app/utils/edit-utils'
import { getFeature } from '../test-utils'
import { EDIT_PENDING_STATUS } from '../../app/constants'

const mockFeature1 = getFeature('node/1')

test('get all retryable edits', () => {
  const edits = [
    {
      'type': 'create',
      'newFeature': mockFeature1,
      'oldFeature': null,
      'comment': 'Test',
      'id': 'node/observe-vfxnxtmo20j',
      'status': EDIT_PENDING_STATUS,
      'errors': [],
      'timestamp': 1562325975841
    },
    {
      'type': 'modify',
      'newFeature': mockFeature1,
      'oldFeature': mockFeature1,
      'comment': 'Test',
      'id': 'node/observe-vfxnxtmo20j',
      'status': EDIT_PENDING_STATUS,
      'errors': [{
        'code': 'VersionMismatchError',
        'extra': 'Version mismatch: Provided 1, server had: 2 of Node 1',
        'message': 'This feature has been updated since you edited it',
        'retryable': false
      }],
      'timestamp': 1562325975841
    },
    {
      'type': 'modify',
      'newFeature': mockFeature1,
      'oldFeature': mockFeature1,
      'comment': 'Test',
      'id': 'node/observe-vfxnxtmo20j',
      'status': EDIT_PENDING_STATUS,
      'errors': [{
        'code': 'UploadNetworkError',
        'extra': '',
        'message': 'Network error while uploading changes',
        'retryable': true
      }],
      'timestamp': 1562325975841
    }
  ]
  expect(getAllRetriable(edits)).toMatchSnapshot()
})
