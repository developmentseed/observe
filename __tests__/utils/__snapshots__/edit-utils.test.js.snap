// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`get all retryable edits 1`] = `
Array [
  Object {
    "comment": "Test",
    "errors": Array [],
    "id": "node/observe-vfxnxtmo20j",
    "newFeature": Object {
      "geometry": Object {
        "coordinates": Array [
          0,
          0,
        ],
        "type": "Point",
      },
      "id": "node/1",
      "properties": Object {
        "id": "node/1",
      },
      "type": "Feature",
    },
    "oldFeature": null,
    "status": "EDIT_PENDING_STATUS",
    "timestamp": 1562325975841,
    "type": "create",
  },
  Object {
    "comment": "Test",
    "errors": Array [
      Object {
        "code": "UploadNetworkError",
        "extra": "",
        "message": "Network error while uploading changes",
        "retryable": true,
      },
    ],
    "id": "node/observe-vfxnxtmo20j",
    "newFeature": Object {
      "geometry": Object {
        "coordinates": Array [
          0,
          0,
        ],
        "type": "Point",
      },
      "id": "node/1",
      "properties": Object {
        "id": "node/1",
      },
      "type": "Feature",
    },
    "oldFeature": Object {
      "geometry": Object {
        "coordinates": Array [
          0,
          0,
        ],
        "type": "Point",
      },
      "id": "node/1",
      "properties": Object {
        "id": "node/1",
      },
      "type": "Feature",
    },
    "status": "EDIT_PENDING_STATUS",
    "timestamp": 1562325975841,
    "type": "modify",
  },
]
`;
